ocaml/xapi/xapi_pusb.ml

```diff
@@ -29,6 +29,60 @@ let create ~__context ~uSB_group ~host ~other_config ~path
   debug "PUSB ref='%s' created" (Ref.string_of pusb);
   pusb
 
+let scan_start ~__context usbs =
+  let host = Helpers.get_localhost ~__context in
+  let known_usb =
+    Db.PUSB.get_all_records ~__context
+    |> List.filter (fun (rf, rc) -> rc.API.pUSB_host = host)
+    |> get_known_usb
+  in
+  let local_usb = get_local_usb usbs in
+  (* Create the newly added pusbs *)
+  USBSet.iter (fun s -> let self = create ~__context ~uSB_group:(Ref.null) ~host ~other_config:[] ~path:s.USB.path ~vendor_id:s.USB.vendor_id
+                   ~vendor_desc:s.USB.vendor_desc ~product_id:s.USB.product_id ~product_desc:s.USB.product_desc ~serial:s.USB.serial
+                   ~version:s.USB.version ~description:s.USB.description in
+               let group = Xapi_pusb_helpers.find_or_create ~__context self in
+               Db.PUSB.set_USB_group ~__context ~self ~value:group
+  ) (USBSet.diff local_usb known_usb);
+
+  List.filter (fun (rf, rc) -> USBSet.mem (extract_known_usb_info rc) (USBSet.diff known_usb local_usb)) (Db.PUSB.get_all_records ~__context)
+  |> List.iter (fun (self, _) ->
+                  let usb_group = Db.PUSB.get_USB_group ~__context ~self in
+                  Db.PUSB.destroy ~__context ~self;
+                  let vusbs = Db.USB_group.get_VUSBs ~__context ~self:usb_group in
+                  List.iter (fun vusb -> Db.VUSB.destroy ~__context ~self:vusb) vusbs;
+                  Db.USB_group.destroy ~__context ~self:usb_group)
+
+let cond = Condition.create ()
+let mutex = Mutex.create ()
+let scan_required : bool ref = ref false
+
+let start_thread f =
+  ignore(Thread.create (fun () ->
+      while true do
+        Mutex.execute mutex (fun () ->
+            while not !scan_required do
+              Condition.wait cond mutex
+            done;
+            (* scan_required is true now. *)
+            scan_required := false;
+          );
+        try
+          f ();
+        with
+          | _ -> failwith "scan failure"
```

Comment:
This will cause the thread to terminate permanently - you need to catch the exception, log it, and throw it away.

Reply:
updated.